<launch>
  <arg name="algorithm"/>
  <arg name="run"/>
  <arg name="record"/>
  <arg name="sim"/>

  <param if="$(arg sim)" name="/use_sim_time" value="true"/>
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find bruce_exploration)/config/rosconsole.config"/>
  <env name="ROSCONSOLE_FORMAT" value="[${severity}] [${time}] [${node}]: ${message}"/>

  <arg name="SLAM_NS" value="/bruce/slam"/>
  <group ns="bruce">
    <group ns="exploration">
      <rosparam command="load" file="$(find bruce_exploration)/config/exploration.yaml"/>

      <node pkg="bruce_exploration" name="server" type="bruce_exploration_server" output="screen">
        <param name="algorithm" value="$(arg algorithm)"/>
        <remap from="~isam2" to="$(arg SLAM_NS)/slam/isam2"/>
        <remap from="~map" to="$(arg SLAM_NS)/mapping/occupancy"/>
        <remap from="~predict_slam_update" to="$(arg SLAM_NS)/slam/predict_slam_update"/>
      </node>

      <node if="$(arg run)" pkg="bruce_exploration" name="client" type="bruce_exploration_client" output="screen">
        <remap from="~costmap" to="/move_base/global_costmap/costmap"/>
        <remap from="~costmap_updates" to="/move_base/global_costmap/costmap_updates"/>
        <remap from="~global_plan" to="/move_base/TebLocalPlannerROS/global_plan"/>
      </node>
    </group>
  </group>

  <include file="$(find bruce_slam)/launch/slam.launch">
    <arg name="rviz" value="false"/>
  </include>

  <include if="$(arg run)" file="$(find bruce_exploration)/launch/move_base.launch">
    <arg name="map_topic" value="$(arg SLAM_NS)/mapping/occupancy"/>
    <arg name="odom_topic" value="$(arg SLAM_NS)/slam/odom"/>
  </include>

  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find bruce_exploration)/config/exploration.rviz"/>

  <node if="$(arg record)" pkg="rosbag" type="record" name="record"
    args="
      -o $(arg algorithm)
      -x '/sonar_oculus_node/image|/bruce/slam/feature_extraction/feature_img|/move_base/(.*)'"
    output="screen"/>

</launch>